2_munge/src/aggregate_inputs.R

```diff
@@ -46,105 +46,81 @@ aggregate_nwis <- function(ind_file, raw_ind_file, remake_file, sites_yml, gd_co
+    time <- convert_time_nc2posix(input_raw$var$streamflow$dim[[2]]) # time value converted to posix
+  } else if (length(dimids) == 3) {
+    streamflow <- matrix(nrow = length(site_inds) * input_raw$dim$time$len, ncol = input_raw$dim$reference_time$len)
+    valid_time <- input_raw$var$streamflow$dim[[2]]$vals
+    ref_time <- convert_time_nc2posix(input_raw$var$streamflow$dim[[3]]) # ref time value converted to posix
+  }
 
-        data.frame(ref_date = rep(date, model_range),
-                   valid_date = valid,
-                   flow = rgamma(model_range,shape = 1, scale = 2),
-                   site_no = rep(sites[site], model_range),
-                   comid = rep(comids_lookup$COMID[comids_lookup$site_id==sites[site]], model_range))
-      }) %>% bind_rows
+  for(s in 1:length(site_inds)) {
+    if(length(dimids) == 2) {
+      streamflow[((s-1)*length(time)+1):((s)*length(time)),1] <- ncvar_get(input_raw, input_raw$var$streamflow,
+                start = c(site_inds[s], 1),
+                count = c(1, -1),
+                raw_datavals = TRUE)
+
+    } else if(length(dimids) == 3) {
+      for(r in 1:input_raw$dim$reference_time$len) {
+        streamflow[((s-1)*length(valid_time)+1):((s)*length(valid_time)),r] <- ncvar_get(input_raw, input_raw$var$streamflow,
+                                                                             start = c(site_inds[s], 1, r),
+                                                                             count = c(1, -1, 1),
+                                                                             raw_datavals = TRUE)
+      }
     }
-    agg_out = rbind(agg_out, agg_cur)
+  }
+
+  #need streamflow, site_no, dateTime, for retro
+  # need streamflow, site_no, refTime, validTime, for forecast
+  if(length(grep('retro',ind_file))>0){
+
+    agg_nwm <- streamflow %>%
+      as.data.frame() %>%
+      setNames('flow') %>%
+      mutate(
+        date = rep(as.Date(time), 3),
+        site_no = rep(comids_lookup$site_id[match(input_raw$dim$feature_id$vals[site_inds],comids_lookup$COMID)], each = length(time))) %>%
+      group_by(site_no, date) %>%
+      summarise(
+        flow = mean(flow)) %>%
+      ungroup()
+
+  }else{
+
+    agg_nwm <- streamflow %>%
+      as.data.frame() %>%
+      setNames(ref_time) %>% # columns are ref_time
+      select(contains('00:00:00')) %>% # we only want ref dates that start at midnight
+      dplyr::do(with(., {
+        ref_dates = rep(as.Date(colnames(.)), each = nrow(.))
+
+        out = data.frame(
+          ref_date = ref_dates,
+          valid_date = rep(valid_time, length(unique(ref_dates)) * length(sites)),
+          flow = as.vector(as.matrix(.)),
+          site_no = rep(rep(comids_lookup$site_id[match(input_raw$dim$feature_id$vals[site_inds],comids_lookup$COMID)],
+                        each = input_raw$dim$time$len), length(unique(ref_dates))))
+      })) # need to figure out what valid date vals mean; what are the time steps?
```

Comment:
sounded from our discussion in the car like the valid_time values are probably number of 3-hour chunks since reference time for medium range (1:80 for 10 days = 8 indices per 24-hr period), number of 6-hour chunks since reference time for long range (180 = 30 days * 6 indices per 24-hr period).

to confirm, those guesses square with my notes from fernando and the descriptions at http://water.noaa.gov/about/nwm